### DISCUSSION TITLE
How to setup a server can authentication with client certificate (mTLS)?

### DISCUSSION THREAD

**Discussion by Leo1003** [salvo-rs/salvo#806] [Q&A]
Currently, I can setup a HTTPS server with client auth enabled in Salvo. However, I cannot find a way to extract the client certificate from `Request` to perform further verification or authorization.

Are there any good example to follow?

**Comment by chrislearn** [comment]
I haven't used mTLS, but I guess you may need to create a ServerConfig that supports mTLS through ConfigBuilder.

https://docs.rs/salvo_core/latest/salvo_core/conn/rustls/struct.ServerConfig.html
https://docs.rs/rustls/0.23.10/rustls/struct.ConfigBuilder.html

**Reply by Leo1003** [reply]
It is important that if we can get the certificate which the incoming request is used in order to do precise permission control.

**Reply by uweber** [reply]
Any luck with extracting the client CN from the TLS connection within salvo? Trying to do the same...
